util/label/label.rs

```diff
@@ -9,49 +9,144 @@ use label_error::LabelError;
 /// TODO: validate . and .. in target name
 /// TODO: validate used characters in target name
 pub fn analyze(input: &'_ str) -> Result<Label<'_>> {
-    let label = input;
-    let (input, repository_name) = consume_repository_name(input, label)?;
-    let (input, package_name) = consume_package_name(input, label)?;
-    let name = consume_name(input, label)?;
-    let name = match (package_name, name) {
-        (None, None) => {
-            return Err(LabelError(err(
-                label,
-                "labels must have a package and/or a name.",
-            )))
+    Label::analyze(input)
+}
+
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone)]
+pub enum Repository<'s> {
+    /// A `@@` prefixed name that is unique within a workspace. E.g. `@@rules_rust~0.1.2~toolchains~local_rustc`
+    Canonical(&'s str), // stringifies to `@@self.0` where `self.0` may be empty
+    /// A `@` (single) prefixed name. E.g. `@rules_rust`.
+    Apparent(&'s str),
+}
+
+impl<'s> Repository<'s> {
+    pub fn repo_name(&self) -> &'s str {
+        match self {
+            Repository::Canonical(name) => &name[2..],
+            Repository::Apparent(name) => &name[1..],
         }
-        (Some(package_name), None) => name_from_package(package_name),
-        (_, Some(name)) => name,
-    };
-    Ok(Label::new(repository_name, package_name, name))
+    }
 }
 
 #[derive(Debug, PartialEq, Eq)]
-pub struct Label<'s> {
-    pub repository_name: Option<&'s str>,
-    pub package_name: Option<&'s str>,
-    pub name: &'s str,
+pub enum Label<'s> {
+    Relative {
+        target_name: &'s str,
+    },
+    Absolute {
+        repository: Option<Repository<'s>>,
+        package_name: Option<&'s str>,
```

Comment:
An absolute label always has a package name (which may happen to be empty), so I don't think this should be an `Option`?

Reply:
Good point. I can update this